mapper/orm/src/main/java/org/hibernate/search/mapper/orm/loading/impl/MutableEntityLoadingOptions.java

```diff
@@ -31,14 +32,19 @@ public void fetchSize(int fetchSize) {
 	}
 
 	public EntityGraphHint<?> entityGraphHintOrNullForType(EntityMappingType entityMappingType) {
-		if ( entityGraphHints == null ) {
+		if ( entityGraphHints == null || entityGraphHints.isEmpty() ) {
 			return null;
 		}
-		String hibernateOrmEntityName = entityMappingType.getEntityName();
-		for ( EntityGraphHint<?> entityGraphHint : entityGraphHints ) {
-			if ( entityGraphHint.graph.appliesTo( hibernateOrmEntityName ) ) {
-				return entityGraphHint;
+		EntityMappingType testedType = entityMappingType;
+		while ( testedType != null ) {
+			for ( EntityGraphHint<?> entityGraphHint : entityGraphHints ) {
+				// This cast is fine because a RootGraph always applies to an entity type
+				EntityDomainType<?> graphedType = (EntityDomainType<?>) entityGraphHint.graph.getGraphedType();
+				if ( graphedType.getHibernateEntityName().equals( testedType.getEntityName() ) ) {
```

Comment:
seems like `graphedType.getTypeName()` might have the same behaviour

Reply:
It seems so, currently... but then the javadoc of `ManagedDomainType#getTypeName` is wrong, since one may customize the (Hibernate ORM) entity name to something completely different than the FQCN, even for non-dynamic-map entities...
I'd rather not rely on that until we clarify.